### ISSUE TITLE
Release code for training stage1, stage2-1, stage 2-2 and  HERMES-p

### ISSUE THREAD

**Issue by oym1994** [LMD0311/HERMES#3]
Nice Job!

I wander when will you also release the code for training stage1, stage2-1, stage 2-2 and  HERMES-p.

Thank you and keep waiting for your kind response!

**Comment by MSunDYY** [comment]
I guess you can just change the config.py to adjust it.

**Comment by LMD0311** [comment]
Thank you for your interest in our work on HERMES!

The training code for stage 1, stage 2-1, and stage 2-2 is already available. You can switch between these stages simply by changing the corresponding configuration file during training.

The code for HERMES-p is quite similar, though it hasn't been fully cleaned up and organized yet. You can, however, try to make some modifications at [hermes_future_render_head.py](https://github.com/LMD0311/HERMES/blob/main/projects/mmdet3d_plugin/models/dense_heads/hermes_future_render_head.py) to adapt it. 

Please also note that for training HERMES-p, we incorporated the VQA data from Omnidrive, which is not used in the standard version. Additionally, the trajectory prediction network is also trained in stage 2-2.

```python
@HEADS.register_module()
class HERMESFutureRenderHead(BEVFormerRenderHead):
    def __init__(self,
                 *args,
                 use_can_bus=True,
                 pred_can_bus=False,
                 **kwargs):
        # ....
        self.use_can_bus = use_can_bus
        self.pred_can_bus = pred_can_bus
        if self.use_can_bus:
            self.ego_feat_mlp = nn.Sequential(
                nn.Linear(9, self.embed_dims * 2),
                nn.LayerNorm(self.embed_dims * 2),
                nn.LeakyReLU(inplace=True),
                nn.Linear(self.embed_dims * 2, self.bev_channels),
                nn.LayerNorm(self.bev_channels),
                nn.LeakyReLU(inplace=True),
            )
            self.ego_hist_traj_mlp = nn.Sequential(
                nn.Linear(2, self.embed_dims * 2),
                nn.LayerNorm(self.embed_dims * 2),
                nn.LeakyReLU(inplace=True),
                nn.Linear(self.embed_dims * 2, self.bev_channels),
                nn.LayerNorm(self.bev_channels),
                nn.LeakyReLU(inplace=True),
            )
            self.ego_command = nn.Parameter(torch.randn(3, self.bev_channels) / (self.bev_channels) ** 0.5)
            xavier_init(self.ego_feat_mlp, distribution='uniform', bias=0.)
            xavier_init(self.ego_hist_traj_mlp, distribution='uniform', bias=0.)

            if self.pred_can_bus:
                self.can_bus_pred_mlp = nn.Sequential(
                    nn.Linear(self.bev_channels, self.embed_dims),
                    nn.LayerNorm(self.embed_dims),
                    nn.LeakyReLU(inplace=True),
                    nn.Linear(self.embed_dims, 4),
                    nn.LayerNorm(4),
                    nn.LeakyReLU(inplace=True),
                    nn.Flatten(1, -1),
                    nn.Linear(4 * self.num_learnable_query, 4),
                )
                xavier_init(self.can_bus_pred_mlp, distribution='uniform', bias=0.)
                
                self.traj_mlp = nn.Sequential(
                    nn.Linear(4, self.embed_dims * 2),
                    nn.LayerNorm(self.embed_dims * 2),
                    nn.LeakyReLU(inplace=True),
                    nn.Linear(self.embed_dims * 2, self.bev_channels),
                    nn.LayerNorm(self.bev_channels),
                    nn.LeakyReLU(inplace=True),
                )
                xavier_init(self.traj_mlp, distribution='uniform', bias=0.)

    def forward(self, ):
        if self.use_downsample:
            if self.num_learnable_query > 0:
                with torch.cuda.amp.autocast():
                    # ....
                    ego_hist_traj = torch.from_numpy(img_metas[0]['ego_hist_traj']).to(
                        input_bev_embed.device).to(torch.float32)[-4:]
                    ego_hist_mask = torch.from_numpy(img_metas[0]['ego_hist_mask']).to(
                        input_bev_embed.device).to(torch.float32)[-4:]
                    ego_hist_vec = ego_hist_traj[1:] - ego_hist_traj[:-1]  # [4, 3]
                    ego_hist_feat = self.ego_hist_traj_mlp(
                        ego_hist_vec[..., :2]) * ego_hist_mask[1:, None]  # [4, 1024]
                    ego_hist_feat = ego_hist_feat.max(dim=0).values
                    ego_feat = torch.from_numpy(img_metas[0]['ego_feat']).to(input_bev_embed.device).to(
                        torch.float32)
                    ego_feat = self.ego_feat_mlp(ego_feat)
                    ego_command = (self.ego_command * torch.from_numpy(img_metas[0]['ego_command']).to(
                        input_bev_embed.device).to(torch.float32)[:, None]).sum(dim=0)
                    ego_feat = ego_feat + ego_hist_feat + ego_command
                    query_embedding = torch.cat(
                        [
                            query_embedding[:,
                            self.num_learnable_query * i: self.num_learnable_query * (i + 1)]
                            + self.frame_embedding[i + 1] + ego_feat
                            for i in range(len_future)
                        ],
                        dim=1).to(self.llm.torch_dtype)

                assert query_embedding.shape[1] == self.num_learnable_query * len_future
                query_embedding = (query_embedding + self.query_pos).to(self.llm.torch_dtype)
            llm_out = checkpoint(self.llm, input_bev_embed, conv, query_embedding, use_reentrant=False)


            llm_bev_embed = llm_out['out_bev']
            if self.training:
                self.loss_chat = llm_out['chat_loss']
            out_query = llm_out['out_query']

            with (torch.cuda.amp.autocast()):
                out_query = out_query.view(-1, self.num_learnable_query, self.bev_channels)
                traj_pred_condition = None

                if self.pred_can_bus:
                    traj_pred = self.can_bus_pred_mlp(
                        out_query.to(torch.float32) + self.frame_embedding[-out_query.shape[0]:, None, :]
                        + ego_feat
                    ).reshape(-1, 2).to(torch.float32)
                    traj_pred = torch.cumsum(traj_pred, dim=0)
                    if self.training:
                        assert 'gt_planning' in img_metas[self.frames[0]].keys()
                        planning_traj = img_metas[self.frames[0]]['gt_planning'][0, :, :2]
                        mask = img_metas[self.frames[0]]['gt_planning_mask'][0].any(axis=1)
                        traj_pred = traj_pred[mask]
                        planning_traj = torch.from_numpy(planning_traj[mask]).to(traj_pred.device).to(torch.float32)

                        self.loss_can_bus = nn.L1Loss(reduction="mean")(planning_traj, traj_pred)
                    assert len(traj_pred) == 6
                    traj_pred_condition = traj_pred.reshape(3, 4)
                    traj_pred_condition = torch.cat(
                        [torch.zeros(1, 4).to(traj_pred_condition.device), traj_pred_condition], dim=0)

                if self.future_generate:
                    assert len(self.frames) > 1
                    bev_embed_down = llm_bev_embed.expand(len(self.frames), -1, -1, -1)
                    if self.use_can_bus and not self.pred_can_bus:
                        bev_embed_down = bev_embed_down + self.frame_embedding[:, None, None, :] + self.can_bus_mlp(
                            torch.cat([torch.from_numpy(img_metas[self.frames[i]]['can_bus']).to(
                                bev_embed_down.device).to(torch.float32).unsqueeze(0)
                                       for i in range(len(self.frames))],
                                      dim=0))[:, None, None, :]
                    else:
                        bev_embed_down = bev_embed_down + self.frame_embedding[:, None, None, :] + self.traj_mlp(
                            traj_pred_condition)[:, None, None, :]

                    bev_embed_down_ = self.down_sample.ego_attention(
                        bev_embed_down[-out_query.shape[0]:, ...].permute(0, 3, 1, 2), out_query)
                    bev_embed_down_current = bev_embed_down[:1]
                    bev_embed_down = torch.cat([bev_embed_down_current, bev_embed_down_], dim=0)

                else:
                    # when not future_generate, the pred_can_bus should be True to avoid unused parameters
                    bev_embed_down = llm_bev_embed

            # uni_feats = checkpoint(self.down_sample.up_sample, bev_embed_down.permute(0, 3, 1, 2))
            uni_feats = self.down_sample.up_sample(bev_embed_down.permute(0, 3, 1, 2))
            # uni_feats = checkpoint(self.bev_upsample, uni_feats)
            # uni_feats = checkpoint(self.render_conv, uni_feats)
            uni_feats = self.render_conv(self.bev_upsample(uni_feats))
            uni_feats_dict = {}
            for i in range(len(self.frames)):
                uni_feats_dict[self.frames[i]] = uni_feats[i]

        else:
            uni_feats = self.get_uni_feats(img_metas, input_bev_embed, only_bev, bs, dtype, fusion_prev=fusion_prev)
            if only_bev:
                return uni_feats

        # ....

```

**Comment by LMD0311** [comment]
> I guess you can just change the config.py to adjust it.

Totally correct! Thank you for your interest!

**Comment by oym1994** [comment]
> Thank you for your interest in our work on HERMES!
> 
> The training code for stage 1, stage 2-1, and stage 2-2 is already available. You can switch between these stages simply by changing the corresponding configuration file during training.
> 
> The code for HERMES-p is quite similar, though it hasn't been fully cleaned up and organized yet. You can, however, try to make some modifications at [hermes_future_render_head.py](https://github.com/LMD0311/HERMES/blob/main/projects/mmdet3d_plugin/models/dense_heads/hermes_future_render_head.py) to adapt it.
> 
> Please also note that for training HERMES-p, we incorporated the VQA data from Omnidrive, which is not used in the standard version. Additionally, the trajectory prediction network is also trained in stage 2-2.
> 
> @HEADS.register_module()
> class HERMESFutureRenderHead(BEVFormerRenderHead):
>     def __init__(self,
>                  *args,
>                  use_can_bus=True,
>                  pred_can_bus=False,
>                  **kwargs):
>         # ....
>         self.use_can_bus = use_can_bus
>         self.pred_can_bus = pred_can_bus
>         if self.use_can_bus:
>             self.ego_feat_mlp = nn.Sequential(
>                 nn.Linear(9, self.embed_dims * 2),
>                 nn.LayerNorm(self.embed_dims * 2),
>                 nn.LeakyReLU(inplace=True),
>                 nn.Linear(self.embed_dims * 2, self.bev_channels),
>                 nn.LayerNorm(self.bev_channels),
>                 nn.LeakyReLU(inplace=True),
>             )
>             self.ego_hist_traj_mlp = nn.Sequential(
>                 nn.Linear(2, self.embed_dims * 2),
>                 nn.LayerNorm(self.embed_dims * 2),
>                 nn.LeakyReLU(inplace=True),
>                 nn.Linear(self.embed_dims * 2, self.bev_channels),
>                 nn.LayerNorm(self.bev_channels),
>                 nn.LeakyReLU(inplace=True),
>             )
>             self.ego_command = nn.Parameter(torch.randn(3, self.bev_channels) / (self.bev_channels) ** 0.5)
>             xavier_init(self.ego_feat_mlp, distribution='uniform', bias=0.)
>             xavier_init(self.ego_hist_traj_mlp, distribution='uniform', bias=0.)
> 
>             if self.pred_can_bus:
>                 self.can_bus_pred_mlp = nn.Sequential(
>                     nn.Linear(self.bev_channels, self.embed_dims),
>                     nn.LayerNorm(self.embed_dims),
>                     nn.LeakyReLU(inplace=True),
>                     nn.Linear(self.embed_dims, 4),
>                     nn.LayerNorm(4),
>                     nn.LeakyReLU(inplace=True),
>                     nn.Flatten(1, -1),
>                     nn.Linear(4 * self.num_learnable_query, 4),
>                 )
>                 xavier_init(self.can_bus_pred_mlp, distribution='uniform', bias=0.)
>                 
>                 self.traj_mlp = nn.Sequential(
>                     nn.Linear(4, self.embed_dims * 2),
>                     nn.LayerNorm(self.embed_dims * 2),
>                     nn.LeakyReLU(inplace=True),
>                     nn.Linear(self.embed_dims * 2, self.bev_channels),
>                     nn.LayerNorm(self.bev_channels),
>                     nn.LeakyReLU(inplace=True),
>                 )
>                 xavier_init(self.traj_mlp, distribution='uniform', bias=0.)
> 
>     def forward(self, ):
>         if self.use_downsample:
>             if self.num_learnable_query > 0:
>                 with torch.cuda.amp.autocast():
>                     # ....
>                     ego_hist_traj = torch.from_numpy(img_metas[0]['ego_hist_traj']).to(
>                         input_bev_embed.device).to(torch.float32)[-4:]
>                     ego_hist_mask = torch.from_numpy(img_metas[0]['ego_hist_mask']).to(
>                         input_bev_embed.device).to(torch.float32)[-4:]
>                     ego_hist_vec = ego_hist_traj[1:] - ego_hist_traj[:-1]  # [4, 3]
>                     ego_hist_feat = self.ego_hist_traj_mlp(
>                         ego_hist_vec[..., :2]) * ego_hist_mask[1:, None]  # [4, 1024]
>                     ego_hist_feat = ego_hist_feat.max(dim=0).values
>                     ego_feat = torch.from_numpy(img_metas[0]['ego_feat']).to(input_bev_embed.device).to(
>                         torch.float32)
>                     ego_feat = self.ego_feat_mlp(ego_feat)
>                     ego_command = (self.ego_command * torch.from_numpy(img_metas[0]['ego_command']).to(
>                         input_bev_embed.device).to(torch.float32)[:, None]).sum(dim=0)
>                     ego_feat = ego_feat + ego_hist_feat + ego_command
>                     query_embedding = torch.cat(
>                         [
>                             query_embedding[:,
>                             self.num_learnable_query * i: self.num_learnable_query * (i + 1)]
>                             + self.frame_embedding[i + 1] + ego_feat
>                             for i in range(len_future)
>                         ],
>                         dim=1).to(self.llm.torch_dtype)
> 
>                 assert query_embedding.shape[1] == self.num_learnable_query * len_future
>                 query_embedding = (query_embedding + self.query_pos).to(self.llm.torch_dtype)
>             llm_out = checkpoint(self.llm, input_bev_embed, conv, query_embedding, use_reentrant=False)
> 
> 
>             llm_bev_embed = llm_out['out_bev']
>             if self.training:
>                 self.loss_chat = llm_out['chat_loss']
>             out_query = llm_out['out_query']
> 
>             with (torch.cuda.amp.autocast()):
>                 out_query = out_query.view(-1, self.num_learnable_query, self.bev_channels)
>                 traj_pred_condition = None
> 
>                 if self.pred_can_bus:
>                     traj_pred = self.can_bus_pred_mlp(
>                         out_query.to(torch.float32) + self.frame_embedding[-out_query.shape[0]:, None, :]
>                         + ego_feat
>                     ).reshape(-1, 2).to(torch.float32)
>                     traj_pred = torch.cumsum(traj_pred, dim=0)
>                     if self.training:
>                         assert 'gt_planning' in img_metas[self.frames[0]].keys()
>                         planning_traj = img_metas[self.frames[0]]['gt_planning'][0, :, :2]
>                         mask = img_metas[self.frames[0]]['gt_planning_mask'][0].any(axis=1)
>                         traj_pred = traj_pred[mask]
>                         planning_traj = torch.from_numpy(planning_traj[mask]).to(traj_pred.device).to(torch.float32)
> 
>                         self.loss_can_bus = nn.L1Loss(reduction="mean")(planning_traj, traj_pred)
>                     assert len(traj_pred) == 6
>                     traj_pred_condition = traj_pred.reshape(3, 4)
>                     traj_pred_condition = torch.cat(
>                         [torch.zeros(1, 4).to(traj_pred_condition.device), traj_pred_condition], dim=0)
> 
>                 if self.future_generate:
>                     assert len(self.frames) > 1
>                     bev_embed_down = llm_bev_embed.expand(len(self.frames), -1, -1, -1)
>                     if self.use_can_bus and not self.pred_can_bus:
>                         bev_embed_down = bev_embed_down + self.frame_embedding[:, None, None, :] + self.can_bus_mlp(
>                             torch.cat([torch.from_numpy(img_metas[self.frames[i]]['can_bus']).to(
>                                 bev_embed_down.device).to(torch.float32).unsqueeze(0)
>                                        for i in range(len(self.frames))],
>                                       dim=0))[:, None, None, :]
>                     else:
>                         bev_embed_down = bev_embed_down + self.frame_embedding[:, None, None, :] + self.traj_mlp(
>                             traj_pred_condition)[:, None, None, :]
> 
>                     bev_embed_down_ = self.down_sample.ego_attention(
>                         bev_embed_down[-out_query.shape[0]:, ...].permute(0, 3, 1, 2), out_query)
>                     bev_embed_down_current = bev_embed_down[:1]
>                     bev_embed_down = torch.cat([bev_embed_down_current, bev_embed_down_], dim=0)
> 
>                 else:
>                     # when not future_generate, the pred_can_bus should be True to avoid unused parameters
>                     bev_embed_down = llm_bev_embed
> 
>             # uni_feats = checkpoint(self.down_sample.up_sample, bev_embed_down.permute(0, 3, 1, 2))
>             uni_feats = self.down_sample.up_sample(bev_embed_down.permute(0, 3, 1, 2))
>             # uni_feats = checkpoint(self.bev_upsample, uni_feats)
>             # uni_feats = checkpoint(self.render_conv, uni_feats)
>             uni_feats = self.render_conv(self.bev_upsample(uni_feats))
>             uni_feats_dict = {}
>             for i in range(len(self.frames)):
>                 uni_feats_dict[self.frames[i]] = uni_feats[i]
> 
>         else:
>             uni_feats = self.get_uni_feats(img_metas, input_bev_embed, only_bev, bs, dtype, fusion_prev=fusion_prev)
>             if only_bev:
>                 return uni_feats
> 
>         # ....

Thank you very much for your kind response! I am a beginner and wish you won't mind these questions!  And I wish to do some advanced work based on this project!
